Add compile-requirements GH workflow and update requirements README (#33126)
openedx/edx-platform: requirements/README.rst

Reviewer: [Nice to have] I added a comment in the other doc, but it would be nice to have a TOC in this doc, even if it comes at the bottom of the very top section.

Author: I don't think GH has an "insert table of contents here" syntax, although files with enough headings automatically get a tiny button that can show the TOC: https://github.blog/changelog/2021-04-13-table-of-contents-support-in-markdown-files/

> .. _OEP-18: https://github.com/openedx/open-edx-proposals/blob/master/oeps/oep-0018-bp-python-dependencies.rst

Reviewer: Nit: I usually keep these references at the bottom of a subsection, but that is just personal preference.

Author: I like keeping them as local as possible unless there are multiple uses of the reference.

> While the ``*.in`` files are intended to be updated manually, the updating the ``*.txt`` files should only be manipulated using Makefile targets in a Linux environment (to match our build and deploy systems). Developers on Mac should prefer to use the GitHub workflows or run Make targets from inside devstack's lms-shell or another Linux environment.

Reviewer: 1. typo fix.
2. Reworded the second sentence to more directly tie it to the former statement.
```suggestion
While the ``*.in`` files are intended to be updated manually, updating the ``*.txt`` files should only be manipulated using Makefile targets in a Linux environment (to match our build and deploy systems). For developers on Mac, this can be achieved by using the GitHub workflows or by running Make targets from inside devstack's lms-shell or another Linux environment.
```

> While the ``*.in`` files are intended to be updated manually, the updating the ``*.txt`` files should only be manipulated using Makefile targets in a Linux environment (to match our build and deploy systems). Developers on Mac should prefer to use the GitHub workflows or run Make targets from inside devstack's lms-shell or another Linux environment.

Author: Note to self: Should tell people to run this on their edx-platform fork if they don't have permissions on the main one.

Workflows and Makefile targets
******************************

Add a dependency
================

> To add a Python dependencies, specify it in the appropriate ``requirements/edx/*.in`` file, push that up to a branch, and then use the `compile-python-requirements.yml workflow <https://github.com/openedx/edx-platform/actions/workflows/compile-python-requirements.yml>`_ to run ``make compile-requirements`` against your branch. This will ensure the lockfiles are updated with any transitive dependencies and will ping you on a PR for updating your branch.

Reviewer: ```suggestion
To add a Python dependency, specify it in the appropriate ``requirements/edx/*.in`` file, push that up to a branch, and then use the `compile-python-requirements.yml workflow <https://github.com/openedx/edx-platform/actions/workflows/compile-python-requirements.yml>`_ to run ``make compile-requirements`` against your branch. This will ensure the lockfiles are updated with any transitive dependencies and will ping you on a PR for updating your branch.
```
